jwst/datamodels/container.py

```diff
@@ -181,9 +187,20 @@ def from_asn(self, asn_data, asn_file_path=None):
         asn_file_path: str
             Filepath of the association, if known.
         """
-        # make a list of all the input files
-        infiles = [member['expname'] for member
-                   in asn_data['products'][0]['members']]
+        # match the allowed_exptypes to the exptype in the association and retain
+        # only those file that match, as a list, if allowed_exptypes is set to none
+        # grab all the files
+        if self.allowed_exptypes:
+            infiles = []
+            logger.debug("Filtering datasets based on allowed exptypes:", self.allowed_exptypes)
+            for member in asn_data['products'][0]['members']:
+                if any([x for x in self.allowed_exptypes if re.match(member['exptype'], x, re.IGNORECASE)]):
```

Comment:
It be good to have `allowed_exptypes` be a string or list of strings.  I.e. if a string is passed in, put it in a list, then use it.  Many of the use cases for it will be to pick out a single exptype, so the call
```
container = datamodels.open('input_asn.json', allowed_exptypes='science')
```
is better than
```
container = datamodels.open('input_asn.json', allowed_exptypes=['science'])
```
I think.

Reply:
But in the case of `calwebb_spec3`, where we need to allow 2 exptypes - science and background - what would the syntax look like? If it's:
```
allowed_exptypes=['science', 'background']
```
then aren't we right back where we started?

Reply:
Yes.  My suggestion is to allow both in `ModelContainer`.  If passed a string, wrap it in a list.

Reply:
I.e the calling sequence should allow
```
container = datamodels.open('input_asn.json', allowed_exptypes='science')
container = datamodels.open('input_asn.json', allowed_exptypes=['science', 'background'])
```

Reply:
On this point, my normal reaction is to "just expect an iterable", don't allow bare `str`, and be done with it. However, because `str` is an iterable, results can be very bizarre, and it will be done often. So, handling the case internally is reasonable.

Reply:
Since this is an internal variable I left it as is.